[LV][EVL] Support in-loop reduction using tail folding with EVL.

Mel-Chen opened the pull request:
Following from #87816, add VPReductionEVLRecipe to describe vector predication reduction.

Address one of TODOs from #76172.

Mel-Chen committed:
Utils support

Mel-Chen committed:
Enable lit test case

Mel-Chen committed:
Remove the constraint for reductions

Mel-Chen committed:
update test case

Mel-Chen committed:
The initial implementation modeled after VPWidenLoadEVLRecipe.

Mel-Chen committed:
Add test case for ordered reduction

Mel-Chen committed:
Add test case for all reduction kinds

Mel-Chen committed:
Block FMinimum and FMaximum

Mel-Chen committed:
remove clone function

fhahn commented on llvm/lib/IR/IRBuilder.cpp:
```diff
@@ -422,6 +436,11 @@ CallInst *IRBuilderBase::CreateFAddReduce(Value *Acc, Value *Src) {
 
+CallInst *IRBuilderBase::CreateFAddReduce(Value *Acc, Value *Src, Value *EVL,
+                                          Value *Mask) {
+  return getReductionIntrinsic(Intrinsic::vp_reduce_fadd, Acc, Src, Mask, EVL);
```
Can we avoid this duplication here by providing a builder that provides an API to create an fadd reduction, which either creates `vector_reduce_fadd` if no mask or EVL has been provided and `vp_reduce_fadd` if mask and EVL has been provided, e.g. by extending `VectorBuilder`?

Then at least the complexity would be unified at a single place.

nikolaypanchenko replied:
We had couple of reasons to duplicate code and extend IRBuilder:
* Less changes in vectorizer and less painful to do pulldown. If VectorBuilder will be used during in VPReductionRecipe, that's not important reason anymore.
* More concrete expectations and not need to handle all variations:
** EVL != nullptr && Mask != nullptr
** EVL != nulltptr && Mask == nullptr
** EVL == nullptr && Mask == nullptr
** EVL == nullptr && Mask != nullptr
Unless there is a plan to support latter, it doesn't look good to pretend that scenario is supported.
BTW, we should really have `&EVL`

Mel-Chen replied:
172745025d81388ef8104006ffd7e6faa3d1b04a
@fhahn @nikolaypanchenko I have attempted to move the utils function of IRBuilderBase to VectorBuilder. Please review and let me know if this design is appropriate. Thank you.

fhahn replied:
The latest version moves the logic to create vector-predication intrinsics to the vector-intrinsic builder, which addressed my main concern (re-use of logic elsewhere), so that looks good to me for now

nikolaypanchenko commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -1598,6 +1628,29 @@ void VPReductionRecipe::print(raw_ostream &O, const Twine &Indent,
          "outside of loop)";
 }
+
+void VPReductionEVLRecipe::print(raw_ostream &O, const Twine &Indent,
```
can you also add vplan test for that recipe ?

Mel-Chen replied:
1de39ab970a3d0094da01a08e68208b77d95224f
For now, I repeated the test of VPReductionRecipe. If there is anything else VPReductionEVLRecipe need to be tested, please let me know.

Mel-Chen committed:
Add VPlan test

Mel-Chen committed:
Move utils to VectorBuilder

Mel-Chen committed:
Revert "Utils support"

This reverts commit 701a4d5e011ae014686b430b3ddf1f12f2a58c88.

alexey-bataev commented on llvm/include/llvm/IR/VectorBuilder.h:
```diff
@@ -92,6 +98,15 @@ class VectorBuilder {
                                  ArrayRef<Value *> VecOpArray,
                                  const Twine &Name = Twine());
+
+  // Emit a VP reduction intrinsic call for recurrence \p Kind.
+  // \p Kind        The kind of recurrence
+  // \p ValTy       The type of operand which the reduction operation is
+  //                performed.
+  // \p VecOpArray  The operand list.
```
```suggestion
  /// Emit a VP reduction intrinsic call for recurrence \p Kind.
  /// \param Kind        The kind of recurrence
  /// \param ValTy       The type of operand which the reduction operation is
  ///                performed.
  /// \param VecOpArray  The operand list.
```

Mel-Chen replied:
2352368231f6ac2d374d76e85eaab543a95265a2
Thanks.

Mel-Chen committed:
Fix comments format

Mel-Chen committed:
Allow null underlyInstr

Mel-Chen committed:
Implement mayWriteToMemory, mayReadFromMemory and mayHaveSideEffects

Mel-Chen committed:
Update llvm/lib/IR/VectorBuilder.cpp

Fix lambda format

Co-authored-by: Alexey Bataev <a.bataev@gmx.com>

fhahn commented on llvm/test/Transforms/LoopVectorize/RISCV/inloop-reduction.ll:
```diff
@@ -1,13 +1,13 @@
 ; RUN: opt -mtriple riscv64-linux-gnu -mattr=+v,+d -passes=loop-vectorize -prefer-inloop-reductions < %s -S -o - | FileCheck %s -check-prefix=INLOOP
-; RUN: opt -passes=loop-vectorize -force-tail-folding-style=data-with-evl -prefer-predicate-over-epilogue=predicate-dont-vectorize -mtriple=riscv64 -mattr=+v -S < %s 2>&1 | FileCheck --check-prefix=IF-EVL %s
+; RUN: opt -passes=loop-vectorize -force-tail-folding-style=data-with-evl -prefer-predicate-over-epilogue=predicate-dont-vectorize -mtriple=riscv64 -mattr=+v -S < %s 2>&1 | FileCheck --check-prefix=IF-EVL-OUTLOOP %s
+; RUN: opt -passes=loop-vectorize -prefer-inloop-reductions -force-tail-folding-style=data-with-evl -prefer-predicate-over-epilogue=predicate-dont-vectorize -mtriple=riscv64 -mattr=+v -S < %s 2>&1 | FileCheck --check-prefix=IF-EVL-INLOOP %s
```
what's the difference between this RUN line and the tests in llvm/test/Transforms/LoopVectorize/RISCV/vectorize-force-tail-with-evl-reduction.ll?

Mel-Chen replied:
It's the same. It's just that `vectorize-force-tail-with-evl-reduction.ll` tests more reduction idioms.

fhahn commented on llvm/test/Transforms/LoopVectorize/RISCV/vectorize-force-tail-with-evl-reduction.ll:
```diff
@@ -0,0 +1,1495 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
+; RUN: opt -passes=loop-vectorize \
```
this only tests in loop reductions, should it be indicated in the test  file name?

Mel-Chen replied:
As I mentioned, if changing the title is possible, we can directly add out-loop reduction tests to this test case.

fhahn replied:
Sure, but the checks for regular reductions are still missing? As the ordered case has been split off, it would be consistent to also split off in loop tests to a separate file, to avoid a single file becoming massive

Mel-Chen replied:
3b527754bfb3a3e27cd729d15f511608e2f7f367
Separated outloop and inloop reduction into two files.

fhahn reviewed (commented):
The title mentions this adds support for in-loop reductions, but I wasn't able to find a check to make sure we only vectorize in-loop reductions?

All tests seem to pass flags guiding towards the use of in-loop/ordered reductions, so the case where the regular reduction strategy is chosen may not be tested well

fhahn commented on llvm/include/llvm/IR/VectorBuilder.h:
```diff
@@ -57,6 +58,11 @@ class VectorBuilder {
     return RetType();
   }
 
+  // Helper function for creating VP intrinsic call.
```
independent of this change, but if `VectorBuilder` only supports generation of vector-predication intrinsics, then it would be better to call it `VectorPredicationBuilder` to avoid confusion

Mel-Chen replied:
Indeed, this can be adjusted later. (But I would prefer a shorter name, perhaps just VectorPredBuilder would be good enough?)

nikolaypanchenko replied:
or `VPBuilder`, like there's `IRBuilder`

fhahn replied:
I think it's worth being more explicit for the name in the utility in `llvm/IR`, `VectorPredBuilder` would sound good to me.

fhahn replied:
nit: use `///` for doc-comment

Mel-Chen commented:
> The title mentions this adds support for in-loop reductions, but I wasn't able to find a check to make sure we only vectorize in-loop reductions?
> 
> All tests seem to pass flags guiding towards the use of in-loop/ordered reductions, so the case where the regular reduction strategy is chosen may not be tested well

Indeed, we might need to change the title.
Originally, I only intended to support in-loop reduction first, but the bad news is that the legality check for folding with EVL occurs before the collection of in-loop reduction. This means we may need to postpone the legality check for folding with EVL or advance the collection of in-loop reduction.
The good news is that I have checked the IR generation in `inloop-reduction.ll` for IF-EVL-OUTLOOP, and it seems to be correct. This means we can directly open up both out-loop and in-loop reduction.
In conclusion, my suggestion is to change the title to "[LV][EVL] Support reduction idioms using tail folding with EVL." and directly add more RUN commands in the test cases to test the results of out-loop reduction. 
@fhahn What do you think?

Mel-Chen commented on llvm/lib/IR/VectorBuilder.cpp:
```diff
@@ -57,7 +57,55 @@ Value *VectorBuilder::createVectorInstruction(unsigned Opcode, Type *ReturnTy,
+    }
+  };
+  auto VPID = getForRecurKind(Kind);
+  if (VPID == Intrinsic::not_intrinsic)
```
Using assert is indeed a more straightforward approach, but it seems like the class intends to support other functionalities as well.
```
enum class Behavior {
    // Abort if the requested VP intrinsic could not be created.
    // This is useful for strict consistency.
    ReportAndAbort = 0,

    // Return a default-initialized value if the requested VP intrinsic could
    // not be created.
    // This is useful for a defensive fallback to non-VP code.
    SilentlyReturnNone = 1,
  };
```
The original author seems to want to support 2 ways, both abort and return nullptr.

fhahn replied:
Right, but here it makes more sense to assert as this is more consistent with the general overall codebase, especially as the caller expects this to always succeed (and not silently return something other than the correct VP intrinsic?)

Mel-Chen replied:
6b3985d764a0a6116ed6f140211b96dac64c3556
Drop support for `SilentlyReturnNone`.

Mel-Chen committed:
Enable vp fminimum and fmaximum

Mel-Chen committed:
Revert "Block FMinimum and FMaximum"

This reverts commit d65e06e4cb477dd088ec462905d0cc593a703c7c.

Mel-Chen committed:
Add assertion for AnyOf

fhahn commented on llvm/lib/Transforms/Utils/LoopUtils.cpp:
```diff
@@ -1192,6 +1192,19 @@ Value *llvm::createSimpleTargetReduction(IRBuilderBase &Builder, Value *Src,
+Value *llvm::createSimpleTargetReduction(VectorBuilder &VBuilder, Value *Src,
+                                         const RecurrenceDescriptor &Desc) {
+  RecurKind Kind = Desc.getRecurrenceKind();
+  assert(!RecurrenceDescriptor::isAnyOfRecurrenceKind(Kind) &&
```
Is this ensured somewhere? Do we have tests with any-of reductions?

Mel-Chen replied:
99081ed139ead25e3a756abdc0f76cc4a3da0398
Added test case for IAnyOf and FAnyOf.

fhahn commented on llvm/test/Transforms/LoopVectorize/RISCV/vectorize-force-tail-with-evl-reduction.ll:
```diff
@@ -0,0 +1,1429 @@
+; IF-EVL-NEXT:    [[TMP17:%.*]] = getelementptr inbounds i32, ptr [[A:%.*]], i64 [[TMP13]]
+; IF-EVL-NEXT:    [[TMP18:%.*]] = getelementptr inbounds i32, ptr [[TMP17]], i32 0
+; IF-EVL-NEXT:    [[VP_OP_LOAD:%.*]] = call <vscale x 4 x i32> @llvm.vp.load.nxv4i32.p0(ptr align 4 [[TMP18]], <vscale x 4 x i1> shufflevector (<vscale x 4 x i1> insertelement (<vscale x 4 x i1> poison, i1 true, i64 0), <vscale x 4 x i1> poison, <vscale x 4 x i32> zeroinitializer), i32 [[TMP12]])
+; IF-EVL-NEXT:    [[TMP19:%.*]] = call i32 @llvm.vp.reduce.umin.nxv4i32(i32 -1, <vscale x 4 x i32> [[VP_OP_LOAD]], <vscale x 4 x i1> [[TMP16]], i32 [[TMP12]])
```
The mask `TMP16` is redundant here correct, as it is handled by EVL, correct?

Could you also add a test case where there is a predicated reduction with a mask different than the header mask?

Mel-Chen replied:
7dc2c50446d9c4e9ad056bd1f4684088e29fcc6f
Good suggestion! These two test cases demonstrate that there is still room for improvement in conditional reduction. Like the redundant select in `@cond_add`, and the issue in `@cond_add_pred` where the header mask and condition cannot be separated.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -1611,6 +1648,29 @@ void VPReductionRecipe::print(raw_ostream &O, const Twine &Indent,
+
+void VPReductionEVLRecipe::print(raw_ostream &O, const Twine &Indent,
+                                 VPSlotTracker &SlotTracker) const {
+  O << Indent << "REDUCE ";
```
Can you add a test for printing the recipe, if there isn't one yet? (Couldn't find one)

Mel-Chen replied:
8f48074fc7b4a8289653fcc6448420443b355a3f
Sure, add the test case for printing the recipe.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -1611,6 +1648,29 @@ void VPReductionRecipe::print(raw_ostream &O, const Twine &Indent,
+  }
+  O << ")";
+  if (RdxDesc.IntermediateStore)
+    O << " (with final reduction value stored in invariant address sank "
```
Do we have a test for the intermediate store case (not necessarily printing, but codegen)?

Mel-Chen replied:
f1c0f9338ca9bca68ba02efb01ca44550c64e375

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1337,23 +1337,29 @@ void VPlanTransforms::addExplicitVectorLength(VPlan &Plan) {
+          llvm_unreachable("unsupported recipe");
+        }
+      } else if (auto *RedR = dyn_cast<VPReductionRecipe>(U)) {
+        auto *N = new VPReductionEVLRecipe(RedR, VPEVL);
```
Shouldn't that also pass `NewMask`, to remove an unnecessary use of the header mask?

Mel-Chen replied:
455b022b51b33632bcc8abb9790ff522d32074ff
It also removed the redundant mask `TMP16` you mentioned, thanks!

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1337,23 +1337,29 @@ void VPlanTransforms::addExplicitVectorLength(VPlan &Plan) {
-      } else {
-        llvm_unreachable("unsupported recipe");
+      if (auto *MemR = dyn_cast<VPWidenMemoryRecipe>(U)) {
+        if (!MemR)
```
this is no op now

Mel-Chen replied:
8b3d67ed15dc7c66591efe5fe8f92837949f2463

fhahn commented:
>The good news is that I have checked the IR generation in inloop-reduction.ll for IF-EVL-OUTLOOP, and it seems to be correct. This means we can directly open up both out-loop and in-loop reduction.
In conclusion, my suggestion is to change the title to "[LV][EVL] Support reduction idioms using tail folding with EVL." and directly add more RUN commands in the test cases to test the results of out-loop reduction.
> @fhahn What do you think?

Sounds good to me, but it would be good to have some upstream buildbot that builds some code with the various options to have some runtime testing.

Mel-Chen committed:
Add test case for anyof

Mel-Chen committed:
Add test case for conditional reduction

Mel-Chen committed:
Add test case for recipe dump

Mel-Chen committed:
Add test case for intermediate store

Mel-Chen committed:
Remove redundant code

Mel-Chen committed:
Split test cases

Mel-Chen committed:
Remove mask(CondOp) if the mask equals header mask.

Mel-Chen committed:
Refine recipe print

Mel-Chen committed:
Replace fatal error with assert.

fhahn commented on llvm/test/Transforms/LoopVectorize/RISCV/vectorize-force-tail-with-evl-cond-reduction.ll:
```diff
@@ -0,0 +1,596 @@
+  %rdx = phi i32 [ %start, %entry ], [ %rdx.add, %for.inc ]
+  %arrayidx = getelementptr inbounds i32, ptr %a, i64 %iv
+  %0 = load i32, ptr %arrayidx, align 4
+  %cmp = icmp sgt i32 %0, 3
```
Could you also add a test with compare of the an induction? I think this may be mis-compiled at the moment, pending https://github.com/llvm/llvm-project/pull/92092?

Mel-Chen replied:
7fbcb005e8402b0030eca349ab4975251298ba56
This is because we cannot guarantee that only the tail iteration EVL might be less than VF\*UF. All operations based on the assumption that "only the tail iteration EVL might be less than VF\*UF" have potential bugs I think.

Mel-Chen committed:
Add conditional reduction test cases with widenInduction.

Mel-Chen committed:
mis-vectctorized after #92092.

fhahn commented on llvm/include/llvm/Transforms/Utils/LoopUtils.h:
```diff
@@ -394,6 +395,8 @@ Value *getShuffleReduction(IRBuilderBase &Builder, Value *Src, unsigned Op,
 /// Fast-math-flags are propagated using the IRBuilder's setting.
 Value *createSimpleTargetReduction(IRBuilderBase &B, Value *Src,
                                    RecurKind RdxKind);
+Value *createSimpleTargetReduction(VectorBuilder &VB, Value *Src,
```
nit: might be good to highlight that this version is used to create vector-predication intrinsics?

Mel-Chen replied:
518051b849bdf0148ff1ec3b499da667873eec0a

fhahn commented on llvm/include/llvm/Transforms/Utils/LoopUtils.h:
```diff
@@ -414,6 +417,9 @@ Value *createTargetReduction(IRBuilderBase &B, const RecurrenceDescriptor &Desc,
 Value *createOrderedReduction(IRBuilderBase &B,
                               const RecurrenceDescriptor &Desc, Value *Src,
                               Value *Start);
+Value *createOrderedReduction(VectorBuilder &VB,
```
nit: might be good to highlight that this version is used to create vector-predication intrinsics?

Mel-Chen replied:
518051b849bdf0148ff1ec3b499da667873eec0a

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1358,23 +1358,30 @@ bool VPlanTransforms::tryAddExplicitVectorLength(VPlan &Plan) {
+        VPValue *NewMask = GetNewMask(MemR->getMask());
+        if (auto *L = dyn_cast<VPWidenLoadRecipe>(MemR)) {
+          auto *N = new VPWidenLoadEVLRecipe(L, VPEVL, NewMask);
+          N->insertBefore(L);
```
would be nice to avoid this duplication here, e.g. by introducing a helper that just returns a the new recipe.

Mel-Chen replied:
05dd9eeae0ebcd5a5da65f7a58b07327f7530672
I plan to extract the recipe replacement into namespace vputils. Maybe like `ReplaceInstWithInst` in BasicBlockUtils.h.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2139,6 +2146,62 @@ class VPReductionRecipe : public VPSingleDefRecipe {
+/// intrinsics, performing a reduction on a vector operand with the explicit
+/// vector length (EVL) into a scalar value, and adding the result to a chain.
+/// The Operands are {ChainOp, VecOp, EVL, [Condition]}.
+class VPReductionEVLRecipe : public VPSingleDefRecipe {
```
I think we need to also add support for the new recipe to type inference in VPlanAnalysis.cpp

Mel-Chen replied:
f2b218320b0d1fcb9bfdeca8d3d8c6072e8057ab

fhahn reviewed (commented):
Thanks for the updates, should be good now in terms of correctness thanks to https://github.com/llvm/llvm-project/pull/92092

Just some final comments left from my side.

Mel-Chen committed:
Refine comments.

Mel-Chen committed:
Refine the code of recipe replacement.

Mel-Chen committed:
Support type inference for VPReductionRecipe and VPReductionEVLRecipe

fhahn commented on llvm/test/Transforms/LoopVectorize/RISCV/vectorize-force-tail-with-evl-reduction.ll:
```diff
@@ -0,0 +1,2253 @@
+; IF-EVL-NEXT:    [[TMP18:%.*]] = getelementptr inbounds i32, ptr [[A:%.*]], i64 [[TMP14]]
+; IF-EVL-NEXT:    [[TMP19:%.*]] = getelementptr inbounds i32, ptr [[TMP18]], i32 0
+; IF-EVL-NEXT:    [[VP_OP_LOAD:%.*]] = call <vscale x 4 x i32> @llvm.vp.load.nxv4i32.p0(ptr align 4 [[TMP19]], <vscale x 4 x i1> shufflevector (<vscale x 4 x i1> insertelement (<vscale x 4 x i1> poison, i1 true, i64 0), <vscale x 4 x i1> poison, <vscale x 4 x i32> zeroinitializer), i32 [[TMP13]])
+; IF-EVL-NEXT:    [[TMP20]] = add <vscale x 4 x i32> [[VP_OP_LOAD]], [[VEC_PHI]]
```
Having another look at this after also taking a look at https://github.com/llvm/llvm-project/pull/93854.

Won't this poison some lanes of the reduction phi in the last (and possibly second-to-last) iteration, as the masked off lanes by the load are poisoned? If that's the case, then it might need to restriction to in-loop reductions, as the per-iteration reduction can make use of EVL of the current iteration.

If this is indeed an issue, I am curious if this would show up in runtime testing and if not it would be interesting how this could be improved?

nikolaypanchenko replied:
Assuming that it's impossible to get wraparound of IV + EVL, Mel's current solution does work right. However, if wraparound is possible such select is a problem and we do really need to emit `vp.merge` to switch to TU policy.
Also, such style of vectorization must avoid use of `ActiveLaneMask` at all costs and it has worse performance impact than properly using runtime VL

Generally, this PR is a good first step to emit vp intrinsics for reductions and follow up PR(s) will introduce `vp.merge` to eliminate use of `ActiveLaneMask`

nikolaypanchenko replied:
@fhahn  I do agree that as-is the changeset is generally unsafe for regular reduction case and the proper fix requires to add another recipe that emits `vp.merge` which later CG transforms into TU. But introduction of that new recipe will complicate the changeset, so how about add a clopt to control reduction vectorization which will be off by default ?  The follow up patch will add the missed recipe to make code correct.

fhahn replied:
IIUC it should be safe for in-loop reductions, right? If that's the case, how about restricting codegen to in-loop reductions for now? this still allows brining up most of the code without miscompiles or the need for a temporary flag.

I am curious if at the moment it would be feasible to catch this issue at runtime, i.e. is there HW that implements EVL in a way that performs the last 2 iterations < VF?

Mel-Chen replied:
I thought it again and realized that we might not need https://github.com/llvm/llvm-project/pull/89580#discussion_r1580754264. The issue probably lies with the mask. I believe 67b3621dbcc8621a2bf9cad265d4213b3018b383 should solve the problem with the second-to-last iteration. The next patch will be performance improvement—transforming the select into a recipe that can generate vp.merge.

Mel-Chen replied:
In-loop reduction might also be unsafe. Please refer to `@cond_add_pred` in `llvm/test/Transforms/LoopVectorize/RISCV/vectorize-force-tail-with-evl-cond-reduction.ll` ([shortcut link](https://github.com/llvm/llvm-project/pull/90184/commits/67b3621dbcc8621a2bf9cad265d4213b3018b383#diff-bb6adf934f6e891dee24590af066f2fb6c5a102051216da8b06aa7f89aa642cfL373)).
I will separate out 67b3621 and open a new pull request.

Mel-Chen replied:
Opened https://github.com/llvm/llvm-project/pull/95368 for the issue of the second-to-last iteration.

fhahn replied:
Ah I see. In that case, to avoid needing to pull in another patch, can we bail out for any operation that use something other than the header mask, so we can land this change, followed by https://github.com/llvm/llvm-project/pull/95368 ?

Mel-Chen replied:
I discussed it with @nikolaypanchenko further and realized my oversight.

The conclusion is that, yes, in-loop reduction is correct and safe, even for examples like `@cond_add_pred`. Previously, I was fixated on the mask needing to be completely correct. However, we only need to ensure that the mask is correct from `lane 0` to `lane evl - 1`. 
Therefore, I will limit this PR to in-loop reduction. Support for out-loop reduction will follow once `VPInstructions::Select (header_mask, true_value, false_value)` can be transformed into `vp.merge`.

fhahn replied:
Sounds good, that matches my understanding as well

nikolaypanchenko commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -1652,6 +1694,29 @@ void VPReductionRecipe::print(raw_ostream &O, const Twine &Indent,
+
+void VPReductionEVLRecipe::print(raw_ostream &O, const Twine &Indent,
+                                 VPSlotTracker &SlotTracker) const {
+  O << Indent << "REDUCE ";
```
Better to have more explicit naming here. For example `VP-REDUCE`

Mel-Chen replied:
Following from other EVL recipes, vp.reduce is already there in the printing of reduction operation.
```
O << " vp.reduce." << Instruction::getOpcodeName(RdxDesc.getOpcode()) << " (";
```

nikolaypanchenko replied:
Since that is a different class, it makes more sense to actually have `VP-REDUCE` as a recipe name  and have `reduce` in rhs

Mel-Chen replied:
I prefer we should temporarily align with the existing EVL recipe. 
Later, we can gather all the developers involved with EVL vectorization to discuss how to standardize the print format for EVL recipes.

fhahn replied:
Sounds good to me as first step, would be good to consistently identify EVL-based recipes in the printing. It would be great if you could share a follow-on PR for that.

Mel-Chen committed:
Introduce ExplicitVectorLengthMask recipe for out-loop reduction.

fhahn commented:
> Sounds good to me, but it would be good to have some upstream buildbot that builds some code with the various options to have some runtime testing.


any more thoughts about runtime testing and why the current runtime testing didn't uncover the issues mentioned?

nikolaypanchenko commented:
> any more thoughts about runtime testing and why the current runtime testing didn't uncover the issues mentioned?

There are few problems to catch such things:
1. At least we don't have HW that supports VL selection not as `vlmax + rest`
2. Upstream QEMU has no option to select other than `vlmax + rest`
3. Our downstream QEMU does have that option and we caught few issues like that, but we don't test upstream LLVM.

@asb is building buildbots and when QEMU option will be available in upstream, that option will be added there too.

Mel-Chen committed:
Revert "Introduce ExplicitVectorLengthMask recipe for out-loop reduction."

This reverts commit 8488520439de8814cfcfa0843f22673c6aabde70.

Mel-Chen committed:
Drop the EVL style vplan because of outloop reduction.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2145,6 +2152,62 @@ class VPReductionRecipe : public VPSingleDefRecipe {
+class VPReductionEVLRecipe : public VPSingleDefRecipe {
+  /// The recurrence decriptor for the reduction in question.
+  const RecurrenceDescriptor &RdxDesc;
+  bool IsOrdered;
```
Is there a reason to not inherit from VPReductionRecipe here?

Mel-Chen replied:
4ffe78d64e7b4c8610448be95fb5724e095248e2
Sure, `VPReductionEVLRecipe` can inherit from `VPReductionRecipe`. I don't have a strong preference on this. However, if possible, would it be better to add a `VPReductionBase` in the future, allowing both `VPReductionRecipe` and `VPReductionEVLRecipe` to inherit from `VPReductionBase`?

Mel-Chen committed:
Inherit VPReductionRecipe.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2102,20 +2103,26 @@ class VPInterleaveRecipe : public VPRecipeBase {
 class VPReductionRecipe : public VPSingleDefRecipe {
+protected:
   /// The recurrence decriptor for the reduction in question.
   const RecurrenceDescriptor &RdxDesc;
```
can those be kept private? Should only be accessed through their accessors?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1411,11 +1411,19 @@ bool VPlanTransforms::tryAddExplicitVectorLength(VPlan &Plan) {
-                isa<VPWidenPointerInductionRecipe>(&Phi));
-      }))
+  bool IncludeWidenInduction = any_of(Header->phis(), [](VPRecipeBase &Phi) {
+    return (isa<VPWidenIntOrFpInductionRecipe>(&Phi) ||
+            isa<VPWidenPointerInductionRecipe>(&Phi));
```
```suggestion
    return isa<VPWidenIntOrFpInductionRecipe, VPWidenPointerInductionRecipe >(&Phi);
```

Mel-Chen replied:
b3b4373ece77e100fff7d99c032932a4e2ee647d

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1411,11 +1411,19 @@ bool VPlanTransforms::tryAddExplicitVectorLength(VPlan &Plan) {
-        return (isa<VPWidenIntOrFpInductionRecipe>(&Phi) ||
-                isa<VPWidenPointerInductionRecipe>(&Phi));
-      }))
+  bool IncludeWidenInduction = any_of(Header->phis(), [](VPRecipeBase &Phi) {
```
nit: `ContainsWideInductions` would be more accurate? Move comment to actual bail out?

Mel-Chen replied:
26fbee2e0af165f4cdcfcd7b4a69f5867484f554
I think the comment should stay where it is. This way, it can be more clearly understood which boolean variable it is referring to.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1411,11 +1411,19 @@ bool VPlanTransforms::tryAddExplicitVectorLength(VPlan &Plan) {
+  });
+  // FIXME: Remove this once we can transform (select header_mask, true_value,
+  // false_value) into vp.merge.
+  bool IncludeOutloopReduction = any_of(Header->phis(), [&](VPRecipeBase &Phi) {
```
`ContainsOutLoopReductions`? Move comment to the bail out?

Mel-Chen replied:
7d2a8c4f77c458389e56bb51c9af83595dea8862
Ditto.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1446,23 +1454,39 @@ bool VPlanTransforms::tryAddExplicitVectorLength(VPlan &Plan) {
-      if (!MemR)
+      VPRecipeBase *NewRecipe = nullptr;
+      auto *CurRecipe = dyn_cast<VPRecipeBase>(U);
+      if (!CurRecipe || CurRecipe->getNumDefinedValues() > 1)
```
why is the NumDefinedValues check needed?

Mel-Chen replied:
Because not all recipes are single def, for example, VPInterleaveRecipe.

fhahn replied:
Right, but why does it matter? Any recipe that isn't handled below will be skipped so there should be no need to check # of defined values unless I am missing something?

Mel-Chen replied:
706c6ed0d2bafba5ff54506f13cbc11c57429896
Make sense. I replaced it with assertion.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1446,23 +1454,39 @@ bool VPlanTransforms::tryAddExplicitVectorLength(VPlan &Plan) {
+        return HeaderMask == OrigMask ? nullptr : OrigMask;
+      };
+      if (auto *MemR = dyn_cast<VPWidenMemoryRecipe>(CurRecipe)) {
+        VPValue *NewMask = GetNewMask(MemR->getMask());
```
Would it be simpler to outline this to a helper function `convertToEVLRecipe` so early returns can be used?

Mel-Chen replied:
I don't quite understand the comment's meaning. Are you referring to [this change](https://github.com/llvm/llvm-project/pull/93854/files#diff-53267225b83e943ceae51c326c9941e323fd7aaf74a08b5e6998d6456f88d1ddR1499)? If so, I think we can just  change this in #93854?

fhahn replied:
Ok sound good!

Mel-Chen committed:
Add virtual and override for getCondOp()

Mel-Chen committed:
Private  RecurrenceDescriptor and ordered bool.

Mel-Chen committed:
Refine WidenInduction check.

Mel-Chen committed:
Rename IncludeWidenInduction to ContainsWidenInductions

Mel-Chen committed:
Rename IncludeOutloopReduction to ContainsOutLoopReductions

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1443,23 +1452,39 @@ bool VPlanTransforms::tryAddExplicitVectorLength(VPlan &Plan) {
+               "New recipe must define the same number of values as the "
+               "original.");
+        NewRecipe->insertBefore(CurRecipe);
+        if (NumDefVal > 0) {
```
As only single-defs are supported (or recipes without users), simpler to dyn_cast to VPSingleDefRecipe here and assert that there are no users in the else case?

Mel-Chen replied:
b59b677c63b01711708f24ff4e4e3698325c40dc
I have handled the load as special case since VPWidenLoadEVLRecipe does not inherit from VPSingleDefRecipe.

fhahn reviewed (commented):
A few more comments inline, but after that I think this should be good to go!

(It looks like a few tests might need updating when updating to latest main)

Mel-Chen committed:
Rebase and update test cases.

Mel-Chen committed:
Remove the check for the number of definations

Mel-Chen committed:
Replace the #definition check with isa<>.

Mel-Chen committed:
Fix the bug from unit tests.

********************
Failed Tests (2):
  LLVM-Unit :: Transforms/Vectorize/./VectorizeTests/21/51
  LLVM-Unit :: Transforms/Vectorize/./VectorizeTests/26/51

Mel-Chen commented:
Ping

Mel-Chen committed:
Refine vectorBuilder

Mel-Chen commented on llvm/lib/IR/VectorBuilder.cpp:
```diff
@@ -57,7 +57,59 @@ Value *VectorBuilder::createVectorInstruction(unsigned Opcode, Type *ReturnTy,
+Value *VectorBuilder::createSimpleTargetReduction(RecurKind Kind, Type *ValTy,
+                                                  ArrayRef<Value *> InstOpArray,
+                                                  const Twine &Name) {
+  auto GetForRecurKind = [](RecurKind Kind) {
```
64f533260fe0b9576e47b7b297a54072a4e3a1a3
And should I drop the lambda too?

fhahn replied:
I guess it would reduce the level of indent and `ID = Intrinsic::vp_reduce_add` is also more compact than `return Intrinsic::vp_reduce_add`

Mel-Chen replied:
2a9352bad457eb43f9fdbe22a347a57bc9940e87
Removed the lambda.

Mel-Chen commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -2144,16 +2151,67 @@ class VPReductionRecipe : public VPSingleDefRecipe {
   VPValue *getVecOp() const { return getOperand(1); }
   /// The VPValue of the condition for the block.
-  VPValue *getCondOp() const {
+  virtual VPValue *getCondOp() const {
```
It's challenging for `VPReduction*`, I think. That's because not all reduction recipes have a `CondOp`. If I put the EVL as *the last* operand, it would be operand(3) if there's a `CondOp`, and operand(2) if there isn't no `CondOp`. Overall, it would be difficult to statically decide whether EVL is operand(2) or operand(3).

fhahn replied:
I don't think there's an easy way around the fact that mask is optional, similar to load/store EVL recipes.

`VPReductionRecipe::classof` likely needs a manual implementation so that isa<VPReductionRecipe>() works for VPReductionEVLRecipe (around like 2143, unfortunately GH doesn't allow commenting on unchanged lines)

Mel-Chen replied:
bdafefa949d5c466007a264c6ff39f8d9098a4c8
62c8979fe81569444ed8a1c4ee51af80295265fe
That's a good reminder.
Although I've refactored some code, to align with the load/store EVL recipes, I'm still keeping the condition as the last operand so far.

Mel-Chen commented:
Ping.

Mel-Chen committed:
Rebase and update test cases.

Mel-Chen committed:
Remove lambda

Mel-Chen committed:
Manually implement classof of VPReductionRecipe.

Mel-Chen committed:
Add isConditional for VPreductionRecipe.

fhahn commented on llvm/include/llvm/IR/VectorBuilder.h:
```diff
@@ -15,6 +15,7 @@
 #ifndef LLVM_IR_VECTORBUILDER_H
 #define LLVM_IR_VECTORBUILDER_H
 
+#include "llvm/Analysis/IVDescriptors.h"
```
nit: use `<>` instead of `""` here for consistency like the other includes below

Mel-Chen replied:
16eb2ab9461d9166cc35d6c52a99f1cc0b540d79

fhahn reviewed (approved):
LGTM, thanks!

Please wait a day or so with committing in case there are any additional thoughts.

Mel-Chen committed:
Unify the format of file include.

Mel-Chen commented:
@fhahn There is an internal compiler error after the rebase. 
I hope #98453 should fix the error.

sanggyu-shin commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1429,11 +1429,20 @@ bool VPlanTransforms::tryAddExplicitVectorLength(VPlan &Plan) {
+  });
+  // FIXME: Remove this once we can transform (select header_mask, true_value,
+  // false_value) into vp.merge.
+  bool ContainsOutloopReductions =
+      any_of(Header->phis(), [&](VPRecipeBase &Phi) {
+        auto *R = dyn_cast<VPReductionPHIRecipe>(&Phi);
+        return R && !R->isInLoop();
+      });
+  if (ContainsWidenInductions || ContainsOutloopReductions)
     return false;
```
Could you explain why loop has out loop reductions can't use EVL ? 
I think EVL can be used even if loop contains out loop reductions.

Mel-Chen replied:
Yes, out-loop reduction for EVL vectorization is feasible, but not at the moment. The reason is that the EVL for the second-to-last iteration might be smaller than VF*UF. This would poison the elements from lane EVL + 1 to lane VF*UF in the merging of the reduction results from the last-to-third and second-to-last iterations.

We will create a patch to address this issue. This patch will transform `select(HeaderMask, LHS, RHS)` and emit `vp.merge`. Once this is done, we can remove the restriction on out-loop reduction.

chapuni commented on llvm/include/llvm/IR/VectorBuilder.h:
```diff
@@ -15,6 +15,7 @@
 #ifndef LLVM_IR_VECTORBUILDER_H
 #define LLVM_IR_VECTORBUILDER_H
 
+#include <llvm/Analysis/IVDescriptors.h>
```
This is a layering violation.

Mel-Chen replied:
Thanks for pointing out this issue. I opened #99276 to fix it.  Please take a look, thanks a lot.

chapuni reviewed (commented):
`llvm/IR` should not depend on `llvm/Analysis`.
